### DISCUSSION TITLE
What is the easiest way to infer a store’s type in TypeScript?

### DISCUSSION THREAD

**Discussion by Zegnat** [ai/nanostores#234] [Q&A] [answered]
Say I have some code that generates a store. The contents of the store can be statically infered. But there is no explicit typing here. Simply:

```typescript
const partOne = {
  a: 15,
  b: "z",
};
const partTwo = {
  c: true,
}
// $ExpectType MapStore<{ a: number; b: string; c: boolean; }>
const $store = map({
  ...partOne,
  ...partTwo,
});
```

Later down the line I would like to use the type of the final Store contents. (I am probably spoiled a little by [Zod’s type inference](https://zod.dev/?id=type-inference).) What would be the (subjectively) best way to infer this? My current solution:

```typescript
type StoreValue<Store extends ReadableAtom> = ReturnType<Store["get"]>;
```

Usage:

```typescript
// $ExpectType { a: number; b: string; c: boolean; }
type myStore = StoreValue<typeof $store>;
```

Am I missing something? Would it be interesting to land an infer helper in the .d.ts files?

**Comment by ai** [comment, marked as answer]
We already have `StoreValue<Store>` type helper in library exports

https://github.com/nanostores/nanostores/blob/main/map/index.d.ts#L23

**Reply by Zegnat** [reply]
And with the same name as my own creation 🤦‍♀️ Thanks for pointing this one out to me!

These helpers are not documented anywhere, are they?

**Reply by ai** [reply]
No, they are mostly small.

But maybe we can create a small list of TS helpers. I will think about it today.

**Reply by ai** [reply]
Added 20ed957 (didn’t find any other useful helpers in the core)

**Reply by Zegnat** [reply]
Thanks!

Even if it saves just one more person from reinventing the wheel, I think this is a win for the README, haha.
